code/modules/paperwork/paper.dm

```diff
@@ -298,56 +298,62 @@
+			var/stamp_class = real_stamp ? sheet.icon_class_name(holding.icon_state) : "FAKE"
+
 			if (isnull(stamps))
 				stamps = list()
-			if(stamps.len < MAX_PAPER_STAMPS)
-				// I hate byond when dealing with freaking lists
-				stamps[++stamps.len] = list(stamp_class, stamp_x, stamp_y, stamp_r)	/// WHHHHY
-
-				/// This does the overlay stuff
-				if (isnull(stamped))
-					stamped = list()
-				if(stamped.len < MAX_PAPER_STAMPS_OVERLAYS)
-					var/mutable_appearance/stampoverlay = mutable_appearance('icons/obj/bureaucracy.dmi', "paper_[stamp_icon_state]")
-					stampoverlay.pixel_x = rand(-2, 2)
-					stampoverlay.pixel_y = rand(-3, 2)
-					add_overlay(stampoverlay)
-					LAZYADD(stamped, stamp_icon_state)
-
-				update_static_data(usr,ui)
-				ui.user.visible_message("<span class='notice'>[ui.user] stamps [src] with [stamp_class]!</span>", "<span class='notice'>You stamp [src] with [stamp_class]!</span>")
-			else
-				to_chat(usr, pick("You try to stamp but you miss!", "There is no where else you can stamp!"))
-			. = TRUE
+
+			// I hate byond when dealing with freaking lists
+			stamps[++stamps.len] = list(stamp_class, stamp_x, stamp_y, stamp_r)	/// WHHHHY
+
+			/// This does the overlay stuff
+			if (isnull(stamped))
+				stamped = list()
+			if(stamped.len < MAX_PAPER_STAMPS_OVERLAYS)
+				var/mutable_appearance/stampoverlay = mutable_appearance('icons/obj/bureaucracy.dmi', "paper_[stamp_icon_state]")
+				stampoverlay.pixel_x = rand(-2, 2)
+				stampoverlay.pixel_y = rand(-3, 2)
+				add_overlay(stampoverlay)
+				LAZYADD(stamped, stamp_icon_state)
+
+			update_static_data(usr,ui)
+			ui.user.visible_message("<span class='notice'>[ui.user] stamps [src] with [stamp_class]!</span>", "<span class='notice'>You stamp [src] with \the [holding.name]!</span>")
 
 		if("save")
 			var/in_paper = params["text"]
 			var/paper_len = length(in_paper)
 			field_counter = params["field_counter"] ? text2num(params["field_counter"]) : field_counter
 
+			if(paper_len == 0)
+				to_chat(ui.user, pick("Writing block strikes again!", "You forgot to write anthing!"))
+				return FALSE
+
 			if(paper_len > MAX_PAPER_LENGTH)
 				// Side note, the only way we should get here is if
 				// the javascript was modified, somehow, outside of
 				// byond.  but right now we are logging it as
 				// the generated html might get beyond this limit
-				log_paper("[key_name(ui.user)] writing to paper [name], and overwrote it by [paper_len-MAX_PAPER_LENGTH]")
-			if(paper_len == 0)
-				to_chat(ui.user, pick("Writing block strikes again!", "You forgot to write anthing!"))
+				log_paper("[key_name(ui.user)] writing to paper [name], and overwrote it by [paper_len - MAX_PAPER_LENGTH] characters (this may be due to internal HTML)")
+				in_paper = copytext(in_paper, 1, MAX_PAPER_LENGTH)
+				to_chat(ui.user, "You run out of room on the paper!")
 			else
 				log_paper("[key_name(ui.user)] writing to paper [name]")
-				if(info != in_paper)
```

Comment:
This check was removed under the assumption that sending an unnecessary TGUI update is cheaper than doing potentially really long string comparisons.